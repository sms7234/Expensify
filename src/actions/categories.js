import database from '../firebase/firebase';

export const setCategories = (categories) => ({
  type:"SET_CATEGORIES",
  categories
});

export const startSetCategories = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`)
      .once('value')
      .then((snapshot) => {
        const categories = [];
        snapshot.forEach((child) => {
          categories.push({
            id: child.key,
            ...child.val()
          });
        });
        dispatch(setCategories(categories));
      });
  };
};

export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) =>{
    const {
      category= '',
      description= ''
    } = categoryData;
    const categoryItem = {category, description};
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`).push(categoryItem).then((ref)=>{
      dispatch(addCategory({
        id: ref.key,
        ...categoryItem
      }));
    });
  };
};

export const editCategory = (id, updates) => ({
  type: 'EDIT_CATEGORY',
  id,
  updates
});

export const startEditCategory = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).update(updates).then(()=>{
      dispatch(editCategory(id,updates));
    });
  };
};

export const removeCategory = (id) => ({
  type: 'REMOVE_CATEGORY',
  id
});

export const startRemoveCategory = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).remove().then(()=>{
      dispatch(removeCategory(id));
    });
  };
};
