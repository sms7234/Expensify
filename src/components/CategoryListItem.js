import React from 'react';
import {Link} from 'react-router-dom';

const CategoryListItem=({dispatch, id, category, description}) =>(
  <Link className="list-item" to={`/add`}>
    <div>
      <h3 className="list-item__title">{category}</h3>
    </div>
    <h3 className="list-item__data">{description}</h3>
  </Link>
)

export default CategoryListItem;
