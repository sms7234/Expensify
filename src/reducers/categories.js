export default (state = [], action) => {
  switch(action.type){
    case 'ADD_CATEGORY':
      return state.concat(action.category);
    case 'SET_CATEGORIES':
      return action.categories;
    default:
      return state;
  }
};
