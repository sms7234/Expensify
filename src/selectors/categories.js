export default (categories, {category}) => {
  return categories.filter((item) => {
    const textMatch = item.category.toLowerCase().includes(category.toLowerCase());
    return textMatch;
  }).sort((a,b)=>{
    if(a.category < b.category){
      return -1;
    } else {
      return 1;
    }
  });
};
