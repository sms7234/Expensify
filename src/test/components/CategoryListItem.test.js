import React from 'react';
import {shallow} from 'enzyme';
import CategoryListItem from '../../components/CategoryListItem';
import categories from '../fixtures/categories';

test('should render CategoryListItem with one item', () => {
  const wrapper = shallow(<CategoryListItem key={categories[0].id}{...categories[0]} />)
  expect(wrapper).toMatchSnapshot();
})
