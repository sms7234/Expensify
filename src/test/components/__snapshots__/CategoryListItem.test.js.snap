// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CategoryListItem with one item 1`] = `
<Link
  className="list-item"
  to="/editCategory/1"
>
  <div>
    <h3
      className="list-item__title"
    >
      Dog
    </h3>
  </div>
  <h3
    className="list-item__data"
  >
    dog costs
  </h3>
</Link>
`;
